Define CCM multi-part API

ronald-cron-arm opened the pull request:
## Description
Define CCM multi-part API along the lines of the GCM multi-part API. The two APIs are not exactly the same as, contrary to GCM, CCM needs the size of the additional data and plaintext/ciphertext from the start.

Fixes #4352

## Status
READY

## Requires Backporting
NO

## Todos
- [x] Documentation
- [x] Changelog updated

mpg commented on include/mbedtls/ccm.h:
```diff
@@ -288,6 +293,141 @@ int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+ *                  to pass successive parts of the input: the plaintext to
+ *                  encrypt, or the ciphertext (not including the tag) to
+ *                  decrypt. After the last part of the input, call
+ *                  mbedtls_ccm_finish().
```
Again, it's probably obvious but perhaps worth saying it explicitly: the total length must be equal to that announced in `starts()`.

mpg commented on include/mbedtls/ccm.h:
```diff
@@ -288,6 +293,141 @@ int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+ *
+ * \return          \c 0 on success.
+ * \return          #MBEDTLS_ERR_CCM_BAD_INPUT on failure:
+ *                  invalid value of \p tag_len.
```
I think there are additional failure cases: the total length of associated data passed to `update_ad()` was less than the value of `ad_len` passed to `starts()`, or the total length of input passed to `update()` was less than announced.

mpg commented on include/mbedtls/ccm.h:
```diff
@@ -288,6 +293,141 @@ int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+ *                  total input length too long.
+ */
+int mbedtls_ccm_update_ad( mbedtls_ccm_context *ctx,
+                           const unsigned char *add,
```
Somewhat pre-existing, so absolutely not a blocker: we're quite inconstent here, sometimes using `ad` (which I think stands for "associated data"), sometimes `add` (which I guess is short for "additional" (data)). Would be better to use the same shorthand everywhere. But again, not a blocker at all.

gilles-peskine-arm replied:
I'd prefer `ad` everywhere to be consistent with AEAD and `update_ad`, and also because “add” makes me think of addition, but not a blocker.

ronald-cron-arm replied:
Changed.

mpg replied:
Ah, just noticed in ChachaPoly we call use `aad` (for additional authenticated data). Sigh.

(No impact on this PR, just venting.)

mpg reviewed (changes requested):
Looks pretty good! I just have a few nitpicks / suggestions for improvement.

mpg commented on include/mbedtls/ccm.h:
```diff
@@ -288,6 +293,141 @@ int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+/**
+ * \brief           This function starts a CCM encryption or decryption
+ *                  operation.
+ *
```
Sorry, I forgot the obvious in my previous review: we need to state explicitly (I think as a doxygen warning) that this function is not implemented in Mbed TLS yet. It will be implemented in a future version, and may be implemented by some `ALT` implementations now, but it's not implemented in Mbed TLS itself right now.

(Same for the other functions declared but not implemented in this PR.)

ronald-cron-arm replied:
roger that, will do.

ronald-cron-arm commented on include/mbedtls/ccm.h:
```diff
@@ -288,6 +293,166 @@ int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+ *
+ * \note            This function is not implemented in Mbed TLS yet.
+ *
+ * \note            For decryption, the output buffer cannot be the same as
```
I didn't get yet a comment about this but I didn't find any checks for that in ccm_auth_crypt(). Not sure what is supported regarding input and output buffer overlap though thus I kept it for additional views on that. @mpg @gilles-peskine-arm do you happen to know something about that?

gilles-peskine-arm replied:
Support for buffer overlap is inconsistent across library modules, often undocumented, and I'm not convinced that all the documentation that is there is actually correct. For PSA we'll need to support arbitrary overlap to comply to the specification, but right now it's a known bug that our PSA crypto implementation does not support overlap.

gilles-peskine-arm replied:
I don't think the 8 bytes constraint makes sense, by the way: I don't see any reason why 8 bytes would matter: pretty much all processing is one block at a time, and a block is 16 bytes for most ciphers. This looks like some bad copypasta from another place, maybe from very old code that only supported ciphers with 8-byte blocks. Fixing this elsewhere is out of scope, but maybe change 8 to 16 in the CCM module?

ronald-cron-arm replied:
Removed.

ronald-cron-arm commented:
@mpg I believe I have addressed your comments. Thanks for having another look.

gilles-peskine-arm commented on include/mbedtls/ccm.h:
```diff
@@ -288,6 +293,166 @@ int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+ *                  of at least \p iv_len Bytes.
+ * \param iv_len    The length of the IV in bytes.
+ * \param total_add_len    The total length of additional data in bytes.
+ * \param total_input_len  The total length of input data to encrypt or decrypt
```
I think it would make more sense to call this the plaintext length (stating explicitly that this excludes the additional data, because “plaintext” can be ambiguous), or the length of the encrypted message (it's both the length what gets encrypted, and the length of the result of the encryption excluding the IV and authentication tag).

gilles-peskine-arm commented on include/mbedtls/ccm.h:
```diff
@@ -288,6 +293,166 @@ int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+ *                  #MBEDTLS_CCM_STAR_DECRYPT.
+ * \param iv        The initialization vector. This must be a readable buffer
+ *                  of at least \p iv_len Bytes.
+ * \param iv_len    The length of the IV in bytes.
```
The permitted values and the consequence on the message length field should be documented, like for the one-shot functions.

gilles-peskine-arm commented on include/mbedtls/ccm.h:
```diff
@@ -288,6 +293,166 @@ int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+ *                         in bytes.
+ *
+ * \return          \c 0 on success.
+ * \return          \#MBEDTLS_ERR_CCM_BAD_INPUT on failure:
```
Or invalid `mode`, or context in a bad state.

gilles-peskine-arm commented on include/mbedtls/ccm.h:
```diff
@@ -288,6 +293,141 @@ int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+ *
+ * \note            This function may be called several times per operation,
+ *                  passing the associated data in chunks.
+ *
```
PSA allows update to fail if there isn't enough AD, but doesn't mandate it, so we're free to decide that we'll only error out in `finish`.

At least, I don't _think_ there are any security implications of leaking partial plaintext/ciphertext if there isn't enough AD, in scenarios where it's ok to leak partial output when there is the expected amount of AD.

ronald-cron-arm replied:
It doesn't seem I have anything to address here.

gilles-peskine-arm replied:
Indeed, I was just confirming that the current behavior is ok.

gilles-peskine-arm commented on include/mbedtls/ccm.h:
```diff
@@ -288,6 +293,166 @@ int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+ *                  ways:
+ *                  - Immediate output: the output length is always equal
+ *                    to the input length.
+ *                  - Buffered output: but for the last part of input data,
```
Minor: in international English, “except” is more widely understood than “but”. Even in proper English, “but” in this sense is rather old-fashioned when used at the beginning of a sentence.

gilles-peskine-arm commented on include/mbedtls/ccm.h:
```diff
@@ -288,6 +293,166 @@ int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+ * \return         \c 0 on success.
+ * \return         #MBEDTLS_ERR_CCM_BAD_INPUT on failure:
+ *                 total input length too long,
+ *                 unsupported input/output buffer overlap detected,
```
I don't think we should mention overlap in the error conditions. We shouldn't promise to detect it: it's difficult.

ronald-cron-arm commented on include/mbedtls/ccm.h:
```diff
@@ -55,6 +55,11 @@
 
+#define MBEDTLS_CCM_DECRYPT       0
+#define MBEDTLS_CCM_ENCRYPT       1
+#define MBEDTLS_CCM_STAR_DECRYPT  2
```
This is to be able to allow in mbedtls_ccm_finish() a tag length of 0 for CCM* but not for CCM as one-shot functions do.

gilles-peskine-arm replied:
Ok. I don't think this is necessary for a low-level API, but it makes sense to do the same as the one-shot functions. (They handle CCM* a bit differently because it was added after the fact.)

ronald-cron-arm commented on ChangeLog.d/m-ccm-api.txt:
```diff
@@ -0,0 +1,5 @@
+Features
+   * The CCM API now supports multi-part operations. Alternative
```
That makes sense. @mpg are you also ok to remove the change log?

mpg replied:
Yes, good idea. Actually I was going to comment on the ChangeLog entry as I was afraid it could confuse people, but I couldn't find a better wording to suggest, so I refrained. Omitting it entirely seems like a good option then :)

But I guess we still want a migration guide entry for the sake of alt implementors, though.

ronald-cron-arm replied:
I am not sure what to say in the migration guide. If they don't want multi-part support I think they don't have to do anything thus there is not really any migration here.

gilles-peskine-arm replied:
Alt implementations must implement the whole module. We don't give them a choice.

ronald-cron-arm replied:
ok then what do they have to do ? Provide empty functions that are not called ?

mpg replied:
My understanding is that they have to provide working functions - unlike us, which makes it very awkward, and is why we'd like to get our own implementation as early as possible (3.1) in order to reduced the window of awkwardness.

mpg replied:
Though in practice until 3.1 they can probably get away with providing empty functions - but that's not future-proof, so I don't think we want to encourage that.

ronald-cron-arm replied:
Provide working functions that are not called? To reduce awkwardness we should then maybe implement the functions with just a call to the alternative implementation ?

ronald-cron-arm replied:
> 
> To reduce awkwardness we should then maybe implement the functions with just a call to the alternative implementation ?
Looking at the code I understand that this question is out of the picture.

mpg replied:
I'm sorry, I don't understand. We can't implement the function with just a call to the alt implementation, because (1) the alt implementation _replaces_ our implementation, and (2) most of the time there is no alt implementation.

I don't see the problem with alt implementations providing functions that are not called - we're a library, we provide functions and have no idea whether someone will call them or not. Same for the alt implementations.

Unless I misunderstood what you wrote, you might be confusing the `ALT` way with how PSA drivers work; `ALT` is much more rudimentary.

mpg replied:
Oh, I replied before I saw your last comment. So I guess it's all clear now.

ronald-cron-arm replied:
ok, I understand better the context now. I should be able to write a migration guide now.

ronald-cron-arm commented:
@mpg @gilles-peskine-arm I believe I have addressed all comments but https://github.com/ARMmbed/mbedtls/pull/4555#discussion_r640364749 for which I am not sure what to do yet.

ronald-cron-arm commented:
@mpg @gilles-peskine-arm  Migration guide added, I believe I've addressed all comments now.

mpg commented on docs/3.0-migration-guide.d/ccm-alt.md:
```diff
@@ -0,0 +1,8 @@
+CCM interface changes: impact for alternative implementations
+-------------------------------------------------------------
+
+The CCM multipart interface has changed with the addition of support for
```
I think this "multipart" should be removed.

ronald-cron-arm replied:
done

mpg reviewed (commented):
Thanks for addressing our feedback, this looks pretty good to me.

However, while working on #4355 I noticed that the API you're proposing here for `starts()` might not be the best to support efficient implementation of the new PSA Crypto multi-part AEAD API, see https://github.com/ARMmbed/mbedtls/issues/4355#issuecomment-849512636 I'd like to resolve that discussion before I'm comfortable approving this PR.

gilles-peskine-arm commented on include/mbedtls/ccm.h:
```diff
@@ -284,10 +289,177 @@ int mbedtls_ccm_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+ *                  \c 13),
+ *                  \p total_ad_len is greater than \c 0xFF00.
+ */
+int mbedtls_ccm_starts( mbedtls_ccm_context *ctx,
```
When looking at CCM in isolation, I liked having `ccm_starts` take both the nonce and the lengths, rather than separate. But [Manuel noticed](https://github.com/ARMmbed/mbedtls/issues/4355#issuecomment-849512636) that this is a poor match for the PSA API which receives the nonce and the lengths in separate functions (for good reasons) and specifies that they can be set in either order (which may have been a bad choice, but it's too late to change now).

So how about removing the lengths from `ccm_starts`, which will align it with `gcm_starts` (but not with chachapoly, which doesn't take a nonce length argument :frowning_face:)? And have a separate function to set the lengths, which can be called after `starts` (it isn't required to support calling it beforehand, but I don't think it would cost anything).

ronald-cron-arm replied:
Yes sure, good that Manuel had an overall look to the AEAD implementation. Otherwise why not aligning the signature of not mbedtls_chachapoly_starts() with that of the CCM and GCM one as well?  Even if we don't do anything with the new nonce length parameter in mbedtls_chachapoly_starts() as we probably don't have the time for that.

gilles-peskine-arm replied:
I think that would be a good idea, but it's not strictly necessary. 3.0 optional material.

ronald-cron-arm replied:
I've added mbedtls_ccm_set_lengths() to the API.

mpg reviewed (dismissed):
LGTM!

gilles-peskine-arm commented on include/mbedtls/ccm.h:
```diff
@@ -316,6 +316,7 @@ int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
  *
  * \return          \c 0 on success.
  * \return          \#MBEDTLS_ERR_CCM_BAD_INPUT on failure:
+ *                  the context has not been initialized,
```
It's impossible to detect that the context has not been initialized: it would require distinguishing whether the current memory content is due to calling `init` followed by other API functions vs uninitialized memory that just happens to contain exactly the same bit values. What we can detect (but might not detect all the time, since this is a low-level use-at-your-own-risk API) is whether the context is in the wrong state, e.g. attempting to call update_ad() after update(), or attempting to call either without starts(). In starts(), an error condition could be “the context has already been started”.

ronald-cron-arm replied:
My bad, in "Improve expected context state" commit, I shouldn't have changed the error code descriptions I guess. I am going to remove them all.

gilles-peskine-arm commented on include/mbedtls/ccm.h:
```diff
@@ -308,6 +308,27 @@ int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
  * \param iv_len    The length of the nonce in Bytes: 7, 8, 9, 10, 11, 12,
  *                  or 13. The length L of the message length field is
  *                  15 - \p iv_len.
+ *
```
In the documentation of starts(), please state that set_lengths() needs to be called before passing any data, and that is can be called before or after starts().

In the documentation of set_lengths(), please state that this function may be called before or after starts(), and must be called before passing any data.

ronald-cron-arm commented:
I've amended commit "Improve expected context state for some APIs" to address https://github.com/ARMmbed/mbedtls/pull/4555#discussion_r642341945 and added a new commit to address https://github.com/ARMmbed/mbedtls/pull/4555#discussion_r642365747.

mpg reviewed (dismissed):
Looks good to me.

gilles-peskine-arm commented on include/mbedtls/ccm.h:
```diff
@@ -284,10 +289,203 @@ int mbedtls_ccm_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+ *                  15 - \p iv_len.
+ *
+ * \return          \c 0 on success.
+ * \return          \#MBEDTLS_ERR_CCM_BAD_INPUT on failure:
```
Here and elsewhere:
```suggestion
 * \return          #MBEDTLS_ERR_CCM_BAD_INPUT on failure:
```

ronald-cron-arm replied:
done

gilles-peskine-arm commented on include/mbedtls/ccm.h:
```diff
@@ -284,10 +289,203 @@ int mbedtls_ccm_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+ *
+ * \return          \c 0 on success.
+ * \return          \#MBEDTLS_ERR_CCM_BAD_INPUT on failure:
+ *                  \p mode is invalid,
```
A context in an invalid state is also a possible reason for `MBEDTLS_ERR_CCM_BAD_INPUT` and we should document it. We just need to be careful not to mandate detecting all possible ways the context can be invalid. We can just say “`ctx` is in an invalid state” without details.

ronald-cron-arm replied:
I thought about that but didn't go that way because:
1) looking at the code of mbedtls_gcm_starts() it does not seem that we want to enforce this in the sense of checking the context state at the beginning of the function. If mbedtls_gcm_starts() is called with a context in an invalid state it will fail (hopefully) with an error code but not sure which one.
2) if we want to enforce that mbedtls_gcm_starts() is called with a context in a valid state then I'd say that MBEDTLS_ERR_CCM_BAD_INPUT is not really a good error code, _BAD_STATE is probably better.

Having said that, looking at the GCM implementation it seems possible for MBEDTLS_ERR_CCM_BAD_INPUT to be returned if the context is in an invalid state thus I can add it if you still prefer. I guess that in that case we would need to add "ctx is in an invalid state" in the documentation of the other new APIs as well, right?

gilles-peskine-arm replied:
A lot of old Mbed TLS documentation has an incomplete description of error conditions. I prefer to be complete in new documentation. We shouldn't promise that all cases of invalid contexts will be detected, but we should document that if the function returns `BAD_INPUT`, it can mean that it's been called with a context in an invalid state.

There's no `BAD_STATE` error in any mbedtls module except chachapoly. Others return `BAD_INPUT_DATA` for a context in a bad state, even high-level modules like cipher.

ronald-cron-arm replied:
done

mpg reviewed (approved):
LGTM!
